modules/auth/sso/sspi_windows.go

```diff
@@ -0,0 +1,303 @@
+func (s *SSPI) Priority() int {
+	return 50000
+}
+
+// VerifyAuthData uses SSPI (Windows implementation of SPNEGO) to authenticate the request.
+// If authentication is successful, returs the corresponding user object.
+// If negotiation should continue or authentication fails, immediately returns a 401 HTTP
+// response code, as required by the SPNEGO protocol.
+func (s *SSPI) VerifyAuthData(ctx *macaron.Context, sess session.Store) *models.User {
+	if !s.shouldAuthenticate(ctx) {
+		return nil
+	}
+
+	cfg, err := s.getConfig()
+	if err != nil {
+		log.Error("could not get SSPI config: %v", err)
+		return nil
+	}
+
+	userInfo, outToken, err := sspiAuth.Authenticate(ctx.Req.Request, ctx.Resp)
+	if err != nil {
+		log.Warn("Authentication failed with error: %v\n", err)
+		sspiAuth.AppendAuthenticateHeader(ctx.Resp, outToken)
+
+		// Include the user login page in the 401 response to allow the user
+		// to login with another authentication method if SSPI authentication
+		// fails
+		addFlashErr(ctx, ctx.Tr("auth.sspi_auth_failed"))
+		ctx.Data["EnableOpenIDSignIn"] = setting.Service.EnableOpenIDSignIn
+		ctx.Data["EnableSSPI"] = true
+		ctx.HTML(401, string(tplSignIn))
+		return nil
+	}
+	if outToken != "" {
+		sspiAuth.AppendAuthenticateHeader(ctx.Resp, outToken)
+	}
+
+	username := sanitizeUsername(userInfo.Username, cfg)
+	log.Info("Authenticated as %s\n", username)
+	if len(username) == 0 {
+		return nil
+	}
+
+	user, err := models.GetUserByName(username)
+	if err != nil {
+		if !models.IsErrUserNotExist(err) {
+			log.Error("GetUserByName: %v", err)
+			return nil
+		}
+		if !cfg.AutoCreateUsers {
+			log.Error("User '%s' not found", username)
+			return nil
+		}
+		user = s.newUser(ctx, username, cfg)
+	}
+
+	// Make sure requests to API paths and PWA resources do not create a new session
+	if !isAPIPath(ctx) && !isAttachmentDownload(ctx) {
+		handleSignIn(ctx, sess, user)
+	}
+
+	return user
+}
+
+// getConfig retrieves the SSPI configuration from login sources
+func (s *SSPI) getConfig() (*models.SSPIConfig, error) {
+	sources, err := models.ActiveLoginSources(models.LoginSSPI)
+	if err != nil {
+		return nil, err
+	}
+	if len(sources) == 0 {
+		return nil, errors.New("no active login sources of type SSPI found")
+	}
+	return sources[0].SSPI(), nil
+}
+
+func (s *SSPI) shouldAuthenticate(ctx *macaron.Context) bool {
+	path := strings.TrimSuffix(ctx.Req.URL.Path, "/")
+	if path == "/user/login" && ctx.Req.FormValue("user_name") != "" && ctx.Req.FormValue("password") != "" {
+		return false
+	} else if ctx.Req.FormValue("auth_with_sspi") == "1" {
+		return true
+	}
+	return !isPublicPage(ctx) && !isPublicResource(ctx)
+}
+
+// newUser creates a new user object for the purpose of automatic registration
+// and populates its name and email with the information present in request headers.
+func (s *SSPI) newUser(ctx *macaron.Context, username string, cfg *models.SSPIConfig) *models.User {
+	email := gouuid.NewV4().String() + "@example.org"
```

Comment:
Let's not use valid domain ;)

```suggestion
	email := gouuid.NewV4().String() + "@localhost"
```

Reply:
Actually `example.org` is a reserved domain for documentation purposes. Quote from [RFC 2606](https://tools.ietf.org/html/rfc2606):

> The Internet Assigned Numbers Authority (IANA) also currently has the following second level domain names reserved which can be used as examples.
>     example.com
>     example.net
>     example.org

I intentionally changed it from `localhost` to a second level domain name, as updating a user profile with email ending in `@location` returns an error that the email is not valid.

Can we keep it as `example.org` or change it to something else like `gitea.local`?

Reply:
In examples yes, this is not really example but user data. We could use `@localhost.localdomain` as that should also resolve to `127.0.0.1`, both in Linux and Windows at least it does.

Reply:
ok, I will use `localhost.localdomain`

Reply:
Actually `localhost.localdomain` is not resolving at Windows 10.

Guess its best to leave it as just `localhost` and fix the email validation in another PR to accept localhost emails?

Reply:
It does for me:
![image](https://user-images.githubusercontent.com/165205/67745544-315a8900-fa2d-11e9-92c1-eecccd3ffb13.png)

Reply:
But anyway even if it does not resolve imho we still can use it

Reply:
ok, using it

Reply:
Should be fixed by [Use loopback name for randomly generated emails](https://github.com/quasoft/gitea/commit/2ce392e7fca9bc139d4a6495e416051dbb5aec31)